network/p2p/gossip/gossip.go

```diff
@@ -231,84 +234,192 @@ func (p *PullGossiper[_]) handleResponse(
+		gossipID := gossipable.GossipID()
+		if !p.set.Has(gossipID) {
+			delete(p.tracking, gossipID)
+			p.discarded.Put(gossipID, nil) // only add to discarded if issued once
+			continue
+		}
+
+		// Ensure we don't attempt to send a gossipable too frequently.
+		lastGossipTime := p.tracking[gossipID]
+		// Invariant: maxRegossipFrequency must be greater than 0 to prevent
+		// gossiping the same transaction multiple times in the same message.
+		if now.Sub(lastGossipTime) < p.maxRegossipFrequency {
+			// Put the gossipable on the front of the queue to keep items sorted
+			// by last issuance time.
+			p.issued.PushLeft(gossipable)
+			break
+		}
+
+		bytes, err := p.marshaller.MarshalGossip(gossipable)
+		if err != nil {
+			// Should never happen because we've already issued this once.
+			delete(p.tracking, gossipID)
+			return err
+		}
+
+		gossip = append(gossip, bytes)
+		sentBytes += len(bytes)
+		p.issued.PushRight(gossipable)
+		p.tracking[gossipID] = now
+	}
+	p.metrics.tracking.Set(float64(len(p.tracking)))
+
+	// Send gossipables to peers
 	msgBytes, err := MarshalAppGossip(gossip)
 	if err != nil {
 		return err
 	}
-
 	sentCountMetric, err := p.metrics.sentCount.GetMetricWith(pushLabels)
 	if err != nil {
 		return fmt.Errorf("failed to get sent count metric: %w", err)
 	}
-
 	sentBytesMetric, err := p.metrics.sentBytes.GetMetricWith(pushLabels)
 	if err != nil {
 		return fmt.Errorf("failed to get sent bytes metric: %w", err)
 	}
-
 	sentCountMetric.Add(float64(len(gossip)))
 	sentBytesMetric.Add(float64(sentBytes))
-
-	return p.client.AppGossip(ctx, msgBytes)
+	if err := p.client.AppGossip(ctx, msgBytes); err != nil {
+		return fmt.Errorf("failed to gossip: %w", err)
+	}
+	return nil
 }
 
-func (p *PushGossiper[T]) Add(gossipables ...T) {
+// Add enqueues new gossipables to be pushed. If a gossiable is already tracked,
+// it is not added again.
+func (p *PushGossiper[T]) Add(ctx context.Context, gossipables ...T) error {
 	p.lock.Lock()
 	defer p.lock.Unlock()
 
+	// Add new gossipables to the pending queue.
+	now := time.Now()
 	for _, gossipable := range gossipables {
-		p.pending.PushRight(gossipable)
+		gossipID := gossipable.GossipID()
+		if _, contains := p.tracking[gossipID]; contains {
+			continue
+		}
+		if _, contains := p.discarded.Get(gossipID); !contains {
+			p.tracking[gossipID] = time.Time{}
+			p.pending.PushRight(gossipable)
+		} else {
+			// Pretend that recently discarded transactions were just gossiped.
+			p.tracking[gossipID] = now
+			p.issued.PushRight(gossipable)
+		}
+	}
+	p.metrics.tracking.Set(float64(len(p.tracking)))
+
+	// If we have too many gossipables, trigger gossip to issue pending
+	// gossipables and/or evict stale entries.
+	//
+	// This invocation of [gossip] may not clear any stale entries (if there are
+	// a bunch of pending, valid gossipables), however, successive calls
+	// eventually will.
+	if p.pending.Len() < p.earlyGossipSize {
```

Comment:
This was previously `len(p.tracking)` rather than `p.pending.Len()`. I think that only looking at `pending` makes more sense to avoid repeatedly calling `gossip` if `earlyGossipSize` is less than the number of txs in the mempool.

Reply:
>if earlyGossipSize is less than the number of txs in the mempool.

I think that case would've still been handled because of the limit on gossip frequency but 🤷